mongo/mongo.go

```diff
@@ -280,8 +280,16 @@ func SelectPeerAddress(addrs []network.Address) string {
 // SelectPeerHostPort returns the HostPort to use as the mongo replica set peer
 // by selecting it from the given hostPorts.
 func SelectPeerHostPort(hostPorts []network.HostPort) string {
+	// TODO(macgreagoir) IPv6. We were always choosing a cloud-local or
+	// localhost addr here (with true arg) but this doesn't make sense in
+	// IPv6, where the IPv6 [public] address is ignored in favour of
+	// ip6-localhost. Only pass true if we find an IPv4 address in the
+	// slice.
 	logger.Debugf("selecting mongo peer hostPort by scope from %+v", hostPorts)
-	return network.SelectMongoHostPortsByScope(hostPorts, true)[0]
+	return network.SelectMongoHostPortsByScope(
+		hostPorts,
+		network.HostPortWithIPv4Address(hostPorts),
```

Comment:
I'm inclined to agree with you about the original, and continuing IPv4, behaviour. Falling back to loopback/localhost can seem like a bad idea for the replicaset config, though it doesn't cause a problem if HA is never enabled.

Can I suggest that this is machine-local question is a separate bug to be considered outside this change, as a new behaviour?